File: linkedlist.py
class Node :
    # __init__() la phuong thuc dac biet sd de tao 1 doi tuong khi no duoc tao
    def __init__(self,data):
        self.data = data # node nay luu tru du lieu
        self.next = None # node nay luu tru con tro giu dia chi node ke tiep

class LinkedList :
    def __init__(self):
        self.head = None # head duoc gan bang None khi chua co Node nao

    def add_data(self,data):
        new_node = Node(data) # khoi tao gia tri cho Node
        if not self.head : # neu self.head chua tro vao dau
            self.head = new_node 
        else :
            temp = self.head 
            while temp.next != None:
                temp = temp.next # cho chay bien temp den node cuoi cung de add them node vao cuoi ll
            temp.next = new_node

    def search_data (self,data):
        temp = self.head
        while temp :
            if temp.data == data :
                return f"{data} exist in linked list"
            temp = temp.next
        return f"{data} not exist in linked list"
    
    def insert_before (self, data, bf_data):
        new_node = Node(data) # khoi tao 1 nut moi
        # neu danh sach rong hoac chen vao dau danh sach
        if not self.head or self.head.data == bf_data :
            new_node.next = self.head # next cua node tro vao head
            self.head = new_node # gan head lai cho new node
        else :
            temp = self.head
            # duyet qua danh sach de tim nut truoc can chen
            while temp.next :
                #temp ke tiep co gia tri = voi bf_data
                if temp.next.data == bf_data:
                    new_node.next = temp.next # lay next cua thang new node tro vao temp ke tiep
                    temp.next = new_node # lay next cua temp hien tai tro vao new node
                    return 
                temp = temp.next

    def delete_node (self,data):
        if not self.head :
            return 
        if self.head.data == data :
            self.head = self.head.next
            return
        
        temp = self.head 
        while temp.next :
            if temp.next.data == data :
                temp.next = temp.next.next
                return 
            temp = temp.next

    
    def traverse (self):
        temp = self.head 
        while temp :
            print(temp.data, end = " ")
            temp = temp.next # di chuyen bien temp sang node ke tiep
        
        print ("")

    def numNodes (self):
        temp = self.head 
        idx = 0 
        while temp :
            idx += 1
            temp = temp.next
        
        return idx

ll = LinkedList()
ll.add_data(1)
ll.add_data(2)
ll.add_data(6)
ll.add_data(4)
ll.add_data(5)
ll.traverse()

print (ll.search_data(4))
ll.insert_before(7,2)
ll.traverse()
ll.delete_node(4)
ll.traverse()
print (ll.numNodes())

File: linkedlist.cpp
#include <iostream>
using namespace std ;

// tao 1 node
struct Node {
    int data ; // thanh phan du lieu
    Node* next ; // thanh phan lien ket
};

Node *createNode (int data){
    Node* node = new Node ; // cap phat dong cho node moi
    node -> data = data ;
    node -> next = nullptr ;
    return node ;
}

struct LinkedList {
    Node* head ;
    Node* tail ;
};

void createNewList (LinkedList &l){
    l.head = nullptr ;
    l.tail = nullptr ;
}

void addToHead (LinkedList &l, Node* node){
    if (l.head == nullptr){
        l.head = node ;
        l.tail = node ;
    }
    else {
        node -> next = l.head ;
        l.head = node ;
    }
}

// add 1 ptu vao dau dslk
void addToTail (LinkedList &l, Node* node){
    if(l.head == nullptr){
        l.head = node ;
        l.tail = node ;
    }
    else {
        l.tail -> next = node ;
        l.tail = node ;
    }
}

// them node p vao sau node q bat ki
void insertAfterQ(LinkedList &l, Node* p, Node* q){
    if (q != nullptr){
        p -> next = q -> next ; // lay next cua p tro vao next ke tiep cua q
        q -> next = p ;
        if (l.tail == p){
            l.tail = p ;
        }
    }
    else {
        addToHead(l,p);
    }
}

// dem so ptu co trong dslk
int numNode (LinkedList l){
    int idx = 0 ;
    Node* node = l.head ;
    while (node != nullptr){
        idx++;
        node = node -> next ;
    }
    return idx ;
}

int removeHead (LinkedList &l, int &x){
    if (l.head != nullptr){
        Node* node = l.head ;
        x = node -> data ; // luu tru gia tri cua node trc khi xoa
        l.head = node -> next ;
        delete node ; // xoa node

        if (l.head == nullptr){
            l.tail = nullptr ;
        }
        return 1 ;
    }
    return 0 ;
}

int removeAfterQ (LinkedList &l, Node* q, int &x){
    if (q != nullptr){
        Node* p = q -> next ;
        if (p != nullptr){
            if (l.tail == p){
                l.tail = q ;
            }
            q -> next = p -> next ;
            x = p -> data ;
            delete p ;
            return 1 ;
        }
        return 0;
    }
    return 0 ;
}

// duyet danh sach va in
void PrintList (LinkedList l){
    if(l.head != nullptr){
        Node* node = l.head ;
        while (node != nullptr){
            cout << node -> data <<" ";
            node = node -> next ;
        }
    }
}

// lay gia tri 1 node bat ki 
Node* getNode (LinkedList &l, int idx){
    Node* node = l.head ;
    int i = 0 ;
    while (node != nullptr && i != idx){
        node = node -> next ;
        i++;
    }
    if (i == idx && node != nullptr){
        return node ;
    }
    return nullptr ;
}

// tim kiem ptu nao do trong linkedlist 
Node* searchIndex (LinkedList &l ,int idx){
    Node* node = l.head ;
    while (node != nullptr && node -> data != idx){
        node = node -> next ;
    }
    if (node != nullptr){
        return node ;
    }
    return nullptr ;
}

// xoa danh sach lien ket 
void destroyList (LinkedList &l){
    int x ;
    Node* node = l.head ;
    while (node != nullptr){
        removeHead(l,x);
        node = l.head ;
    }
    l.tail = nullptr ;
}

int main () {
    // create a linked list 
    LinkedList l ;
    createNewList(l);

    // add another data 
    Node* node ;
    // srand((int)time(0));
    for (int i = 0 ; i < 10 ; i++){
        // int r = 20 + rand() % (20 + 1 - 1);
        node = createNode(i);
        addToTail(l,node);
    }

    // print list 
    PrintList(l);
    cout << endl ;

    // length of linkedlist 
    int len = numNode(l);
    cout << len << endl ;

    // get index at idx = ...
    Node* atidx7 = getNode(l,7);
    if (getNode(l,7) != nullptr){
        cout << atidx7 -> data << endl ;
    }

    // search 4 in list 
    Node* search4 = searchIndex(l,4);
    if (search4 != nullptr){
        cout <<"4 in here";
    }
    else {
        cout <<"4 not in here";
    }

    cout << endl ;

    // insert 1106 after 4 ;
    Node* node1106 = createNode(1106);
    insertAfterQ(l,node1106,search4);
    PrintList(l);

    while (l.head != nullptr){
        Node* temp = l.head ;
        l.head = l.head -> next ;
        delete temp ;
    }

    return 0 ;
}
